File: app/src/main/java/ru/anasoft/calculator/Action.java
package ru.anasoft.calculator;

public class Action {

    private StringBuilder inputParam;
    private double param;
    private String lastOperation;
    private double result;
    private boolean usePoint;

    public Action() {
        this.inputParam = new StringBuilder("0");
        clearParam();
        clearResult();
    }

    public void clearParam() {
        this.param = 0;
        this.inputParam.replace(0, this.inputParam.length(), "0");
        this.usePoint = false;
    }

    public void clearResult() {
        this.lastOperation = "=";
        this.result = 0;
    }

    public void addNumToParam(String num) {
        if (this.inputParam.toString().equals("0")) {
            this.inputParam.delete(0, 1);
        }
        this.inputParam.append(num);
        this.param = Double.parseDouble(this.inputParam.toString());
    }

    public String getParam() {
        return this.inputParam.toString();
    }

    public String getResult() {
        return String.valueOf(this.result);
    }

    public void doOperation(String operation) {

        switch (this.lastOperation){
            case ("+"):
                this.result = this.result + this.param;
                break;
            case ("-"):
                this.result = this.result - this.param;
                break;
            case ("*"):
                this.result = this.result * this.param;
                break;
            case ("/"):
                this.result = this.result / this.param;
                break;
            case ("="):
                if (this.param != 0) {
                    this.result = this.param;
                }
                break;
        }

        this.lastOperation = operation;
        clearParam();

    }

    public void addPoint() {
        if (!this.usePoint) {
            this.usePoint = true;

            if (this.inputParam.length() == 0) {
                this.inputParam.append("0");
            }
            this.inputParam.append(".");
        }
    }
}

File: app/src/main/java/ru/anasoft/calculator/Action.kt
package ru.anasoft.calculator

import java.lang.StringBuilder

class Action {
    private val inputParam: StringBuilder = StringBuilder("0")
    private var param = 0.0
    private var lastOperation: String? = null
    private var result = 0.0
    private var usePoint = false
    fun clearParam() {
        param = 0.0
        inputParam.replace(0, inputParam.length, "0")
        usePoint = false
    }

    fun clearResult() {
        lastOperation = "="
        result = 0.0
    }

    fun addNumToParam(num: String?) {
        if (inputParam.toString() == "0") {
            inputParam.delete(0, 1)
        }
        inputParam.append(num)
        param = inputParam.toString().toDouble()
    }

    fun getParam(): String {
        return inputParam.toString()
    }

    fun getResult(): String {
        return result.toString()
    }

    fun doOperation(operation: String?) {
        when (lastOperation) {
            "+" -> result += param
            "-" -> result -= param
            "*" -> result *= param
            "/" -> result /= param
            "=" -> if (param != 0.0) {
                result = param
            }
        }
        lastOperation = operation
        clearParam()
    }

    fun addPoint() {
        if (!usePoint) {
            usePoint = true
            inputParam.append(".")
        }
    }

    init {
        clearParam()
        clearResult()
    }
}